File: src/types/_Enhancements.ts
/**
 * @title Enhancements
 */

import {
  ChildEntities,
  Entity,
  IncludeIdentifier,
  Integer,
  Object,
  Optional,
  Required,
  String,
  TypeAlias,
} from "tsondb/schema/dsl"
import { SkillWithEnhancementsIdentifier } from "./_IdentifierGroup.js"
import { EnhancementPrerequisites } from "./_Prerequisite.js"
import { NestedTranslationMap } from "./Locale.js"
import { Errata } from "./source/_Erratum.js"
import { PublicationRefs } from "./source/_PublicationRef.js"

export const Enhancements = TypeAlias(import.meta.url, {
  name: "Enhancements",
  comment: "A list of enhancements.",
  type: () => ChildEntities(Enhancement),
})

export const Enhancement = Entity(import.meta.url, {
  name: "Enhancement",
  namePlural: "Enhancements",
  type: () =>
    Object({
      parent: Required({
        comment: "The spellwork or liturgy this enhancement belongs to",
        type: IncludeIdentifier(SkillWithEnhancementsIdentifier),
      }),
      skill_rating: Required({
        comment: "The skill rating required to learn this enhancement.",
        type: Integer({
          minimum: 8,
          maximum: 16,
          multipleOf: 2,
        }),
      }),
      adventure_points_modifier: Required({
        comment:
          "The value to multiply with the numeric representation of the associated skill's improvement cost to form the final AP cost of this enhancement.",
        type: Integer({
          minimum: 1,
        }),
      }),
      prerequisites: Optional({
        type: IncludeIdentifier(EnhancementPrerequisites),
      }),
      src: Optional({
        comment: "Only defined if different than the associated skill.",
        type: IncludeIdentifier(PublicationRefs),
      }),
      translations: NestedTranslationMap(
        Required,
        "Enhancement",
        Object({
          name: Required({
            comment: "The name of the enhancement.",
            type: String({
              minLength: 1,
            }),
          }),
          effect: Required({
            comment: "The effect description.",
            type: String({
              minLength: 1,
              isMarkdown: true,
            }),
          }),
          errata: Optional({
            type: IncludeIdentifier(Errata),
          }),
        }),
      ),
    }),
  parentReferenceKey: "parent",
  instanceDisplayName: {},
  instanceDisplayNameCustomizer: ({
    instance,
    instanceDisplayName,
    instanceDisplayNameLocaleId,
    getDisplayNameForInstanceId,
  }) => ({
    name: `${getDisplayNameForInstanceId(instance.parent)?.name ?? ""} — ${instanceDisplayName}`,
    localeId: instanceDisplayNameLocaleId,
  }),
  uniqueConstraints: [
    [
      {
        keyPath: "parent",
      },
      {
        entityMapKeyPath: "translations",
        keyPathInEntityMap: "name",
      },
    ],
  ],
})

File: Sources/OptolithDatabaseSchema/Types/_Enhancements.swift
import FileDB

/// A list of enhancements.
@TypeAlias
public struct Enhancements {
  @MinItems(3)
  @MaxItems(5)
  @Relationship(Enhancement.self)
  let list: [Enhancement.ID]
}

@Model
public struct Enhancement {

      /// The skill rating required to learn this enhancement.
      @Minimum(8)
      @Maximum(16)
      @MultipleOf(2)
      let skill_rating: Int
      /// The value to multiply with the numeric representation of the associated skill's improvement cost to form the final AP cost of this enhancement.
      @Minimum(1)
      let adventure_points_modifier: Int

      let prerequisites: EnhancementPrerequisites?

  /// Only defined if different than the associated skill.
  @MinItems(1)
  let src: [PublicationRef]?

    /// All translations for the entry, identified by IETF language tag (BCP47).
    @Relationship(Locale.self)
    let translations: [String: Translation]

    @Embedded
    struct Translation { // EnhancementTranslation
          /// The name of the enhancement.
          @MinLength(1)
          let name: String

          /// The effect description.
          @MinLength(1)
          @Markdown
          let effect: String

        /// A list of errata for the entry in the specific language.
        @MinItems(1)
        let errata: [Erratum]?
    }
}
